From 46f24df22ecff5668d95b0c1add7f8c159f3d4fd Mon Sep 17 00:00:00 2001
From: nagaraju <nmekala@xilix.com>
Date: Tue, 8 Oct 2013 11:02:06 +0530
Subject: [PATCH 26/28] [Patch, microblaze]: Correct insn pattern definition
 for lshrsi3_with_size_opt

rtl expression for lshrsi3_with_size_opt should be lshiftrt.
Change logic so if statement uses a addk to shift into operand 0 if required.
Use srl to shift right in the branch delay slot.

Move pattern down to appropriate location in logical shift right group of the
machine description for microblaze - after the use of barrel shifts if available,
but before the default action for shifting based on an immediate where a loop
of srl's are emitted.

Signed-off-by:Nagaraju <nmekala@xilix.com>
Signed-off-by: David Holsgrove <david.holsgrove@xilinx.com>
Signed-off-by: Stephan Linz <linz@li-pro.net>
---
 gcc/config/microblaze/microblaze.md | 42 ++++++++++++++++++-------------------
 1 file changed, 21 insertions(+), 21 deletions(-)

diff --git a/gcc/config/microblaze/microblaze.md b/gcc/config/microblaze/microblaze.md
index baf8e6d..0cf1960 100644
--- a/gcc/config/microblaze/microblaze.md
+++ b/gcc/config/microblaze/microblaze.md
@@ -1378,27 +1378,6 @@
   (set_attr "length"   "80")]
 )
 
-(define_insn "*lshrsi3_with_size_opt"
-  [(set (match_operand:SI 0 "register_operand" "=&d")
-       (ashift:SI (match_operand:SI 1 "register_operand"  "d")
-                   (match_operand:SI 2 "immediate_operand" "I")))]
-  "(INTVAL (operands[2]) > 5 && optimize_size)"
-  {
-    operands[3] = gen_rtx_REG (SImode, MB_ABI_ASM_TEMP_REGNUM);
-
-    output_asm_insn ("ori\t%3,r0,%2", operands);
-    if (REGNO (operands[0]) != REGNO (operands[1]))
-        output_asm_insn ("srl\t%0,%1", operands);
-
-    output_asm_insn ("addik\t%3,%3,-1", operands);
-    output_asm_insn ("bneid\t%3,.-4", operands);
-    return "addk\t%0,%0,%0";
-  }
-  [(set_attr "type"    "multi")
-  (set_attr "mode"    "SI")
-  (set_attr "length"  "20")]
-)
-
 (define_insn "*ashlsi_inline"
   [(set (match_operand:SI 0 "register_operand" "=&d")
        (ashift:SI (match_operand:SI 1 "register_operand"  "d")
@@ -1577,6 +1556,27 @@
   (set_attr "length"	"4,4")]
 )
 
+(define_insn "*lshrsi3_with_size_opt"
+  [(set (match_operand:SI 0 "register_operand" "=&d")
+       (lshiftrt:SI (match_operand:SI 1 "register_operand"  "d")
+                   (match_operand:SI 2 "immediate_operand" "I")))]
+  "(INTVAL (operands[2]) > 5 && optimize_size)"
+  {
+    operands[3] = gen_rtx_REG (SImode, MB_ABI_ASM_TEMP_REGNUM);
+
+    output_asm_insn ("ori\t%3,r0,%2", operands);
+    if (REGNO (operands[0]) != REGNO (operands[1]))
+        output_asm_insn ("addk\t%0,%1,r0", operands);
+
+    output_asm_insn ("addik\t%3,%3,-1", operands);
+    output_asm_insn ("bneid\t%3,.-4", operands);
+    return "srl\t%0,%0";
+  }
+  [(set_attr "type"    "multi")
+  (set_attr "mode"    "SI")
+  (set_attr "length"  "20")]
+)
+
 (define_insn "*lshrsi_inline"
   [(set (match_operand:SI 0 "register_operand" "=&d")
        (lshiftrt:SI (match_operand:SI 1 "register_operand"  "d")
-- 
1.8.3.4

